custom/src/main/java/co/elastic/otel/config/DynamicInstrumentation.java

```diff
@@ -0,0 +1,261 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.otel.config;
+
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.api.trace.TracerProvider;
+import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
+import io.opentelemetry.sdk.internal.ComponentRegistry;
+import io.opentelemetry.sdk.internal.ScopeConfigurator;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
+import io.opentelemetry.sdk.trace.SdkTracerProviderBuilder;
+import io.opentelemetry.sdk.trace.internal.TracerConfig;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Notes: 1. The instrumentation can't have been disabled by configuration, eg using
+ * -Dotel.instrumentation.[name].enabled=false as in that case it is never initialized so can't be
+ * "re-enabled" 2. The specific instrumentation name is used, you can see these by setting this
+ * class logging level to j.u.l.Level.CONFIG 3. The disable/re-enable is eventually consistent,
+ * needing the application to pass a synchronization barrier to take effect - but for most
+ * applications these are very frequent
+ */
+public class DynamicInstrumentation {
+
+  private static final Logger logger = Logger.getLogger(DynamicInstrumentation.class.getName());
+  public static final String INSTRUMENTATION_NAME_PREPEND = "io.opentelemetry.";
+  // note the option can't be an env because no OSes support changing envs while the program runs
+  public static final String INSTRUMENTATION_DISABLE_OPTION =
+      "elastic.otel.java.disable_instrumentations";
+
+  private static Object getField(String fieldname, Object target) {
+    try {
+      Field field = target.getClass().getDeclaredField(fieldname);
+      field.setAccessible(true);
+      return field.get(target);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      throw new IllegalStateException(
+          "Error getting " + fieldname + " from " + target.getClass(), e);
+    }
+  }
+
+  private static Object call(String methodname, Object target) {
+    try {
+      Method method = target.getClass().getDeclaredMethod(methodname);
+      method.setAccessible(true);
+      return method.invoke(target);
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+      throw new IllegalStateException(
+          "Error calling " + methodname + " on " + target.getClass(), e);
+    }
+  }
+
+  private static Object call(String methodname, Object target, Object arg1, Class<?> arg1Class) {
```

Comment:
```suggestion
  private static <T> Object call(String methodname, Object target, T arg1, Class<? super T> arg1Class) {
```

[Nit] introduces a minimal bit of type safety